crypto/ec/ec_key.c

```diff
@@ -517,6 +519,34 @@ int ec_key_private_check(const EC_KEY *eckey)
     return 1;
 }
 
+#ifndef OPENSSL_NO_SM2
+int sm2_key_private_check(const EC_KEY *eckey)
```

Comment:
This feels hugely misplaced.  Create a new file in `crypto/sm2`, `sm2_key.c` perhaps?

Reply:
That was my first idea as well, but then I realized that SM2 keygen is still happening in this file and it felt wrong to put validation somewhere else!

Would you insist on splitting validation in its own file?

Reply:
This is specifically an *sm2* function...  does that clarify how I think?  The key generation is combined in one function that looks at the internal key type (or more precisely, the flag `EC_FLAG_SM2_RANGE`) as indication how it should generate the key.

You *could* choose to include the SM2 validation in `ec_key_private_check()`.  That would be much more acceptable in this file, in my view.

Choose.

Reply:
I compromised with an `sm2_key.c` file, but in the `crypto/ec` folder, as we don't have low level SM2 objects so `sm2_key_private_check()` feels to me like an `EC_KEY` operation (that is what it operates on).

See 30ba1329ed8d6d11c1ad75728f50b87b2f38aba2

Reply:
Er, files in `crypto/sm2/` operate on `EC_KEY` objects as well, but `crypto/sm2/` is already consecrated SM2 specific functions (in terms of *functionality*, not in terms of types)...  having one such file (and function) in `crypto/ec/` breaks the existing file organization.  It comes off as a bit jarring, related to already existing stuff.

If that makes you feel better, everything in `crypto/sm2/` could be moved to `crypto/ec/`, since we've already piled everything else that's EC related there (such as Ed25519 et al).

Reply:
You got me! I completely erased from my mind that a `crypto/sm2` folder already existed and is populated with similar SM2 specific uses of `EC_KEY` objects, I agree with you and will move it in a upcoming PR!

Thanks for your insistence!!

Reply:
See 2d233c372251f541c119210a1c3c136d83454720

I am not particularly proud of

```C
#include "../ec/ec_local.h"
```

but I feel internal resistance to the idea of rewriting it to use the getters: if `SM2_KEY` existed we would do direct access in this kind of low level function, as we do in the `ec_key_private_check()` function from which the SM2 specific variant is derived.

So unless @levitte or another reviewer feels strongly about avoiding this practice for this specific case, I would keep it.
My rationale, even if in general I dislike the practice of including internal headers from other submodules, is that the `crypto/sm2` case is special in being a _conceptual_ grouping (which is intentionally not matched with dedicated data types for the submodule), and future functions belonging in `sm2_key.c` would likely share with `sm2_key_private_check()` the need to access the fields of `EC_KEY` data objects at the lowest level of abstraction.

Reply:
I would like to see you use the getters, just like the rest of the `crypto/sm2/` code.

Reply:
Addressed with b5b1407ff7650db5641a0b8855177f3ef6df413a